drivers/flash/flash_sam0.c

```diff
@@ -301,12 +311,21 @@ static int flash_sam0_write(const struct device *dev, off_t offset,
 
 	err = flash_sam0_write_protection(dev, false);
 	if (err == 0) {
-		for (idx = 0; idx < len; idx += FLASH_PAGE_SIZE) {
-			err = flash_sam0_write_page(dev, offset + idx,
-						    &pdata[idx]);
+		while (len > 0) {
+			size_t eop_len, write_len;
+
+			/* Maximum size without crossing a page */
+			eop_len = FLASH_PAGE_SIZE - (offset % FLASH_PAGE_SIZE);
```

Comment:
This part is important only for the first loop run so it can be as well executed outside of loop, before entering it for the first time; when moved, the `eop_len` can just become pre-initialized `write_len`, that will be corrected, when needed, by line 319.

Reply:
@de-nordic, say you were writing FLASH_PAGE_SIZE + 8 bytes.  If that is the case, and since we increment offset, you would want to calculate eop_len on every loop?  If you change eop_len to FLASH_PAGE_SIZE then what you stated would work (which may be what you are suggesting)?

Reply:
No. As far as I understand you predict that somebody will try to write, let say, `len > FLASH_PAGE_SIZE`, which means that you have full page + maybe prefix of bytes and maybe suffix of bytes. Something like 
`start_bytes + N * FLASH_PAGE_SIZE + end_bytes`, where `0<= start_bytes < FLASH_PAGE_SIZE`, `0 <= end_bytes  < FLASH_PAGE_SIZE` and N is [0, whatever].
You can not cross page boundary so you have to write `start_bytes` first, then you can have  N page aligned chunks  and then write `end_bytes`.
So your first `write_len` would be `start_bytes`, and you can calculate it before entering the loop; then you are aligned and can write `write_len == FLASH_PAGE_SIZE` N times and finally write `end_bytes`.

So within loop, after you update `len -= write_len`, as you do now, you can just update `write_len = MIN(FLASH_PAGE_SIZE, len)`.
With the simple change you remove some of calculations out of the loop.
But it is Friday and I may have got that wrong.

Reply:
I hear you about Friday.  I think I get the logic now.  I pushed an updated version of the code...